test/jest/Bladeburner/Actions.test.ts

```diff
@@ -0,0 +1,358 @@
+        /** Testing against a guaranteed 0-chaos level of charisma */
+        basicStats(), Player.gainCharismaExp(1e500), allCitiesHighChaos(), complete(Diplomacy);
+        expect(inst.cities[otherCity].chaos).toBeGreaterThan(0);
+      });
+    });
+
+    describe(BladeburnerGeneralActionName.FieldAnalysis, () => {
+      const Field = GeneralAction.createId(BladeburnerGeneralActionName.FieldAnalysis);
+
+      it("improves population estimate", () => {
+        basicStats(),
+          ({ pop, popEst: before } = inst.getCurrentCity()),
+          complete(Field),
+          finish(),
+          ({ popEst: after } = inst.getCurrentCity());
+        expect(Math.abs(after - pop)).toBeLessThan(Math.abs(before - pop));
+      });
+
+      it.each(<(keyof Skills)[]>["hacking", "charisma"])("awards %s exp", (stat: keyof Skills) => {
+        basicStats(), (before = Player.exp[stat]), complete(Field);
+        expect(Player.exp[stat]).toBeGreaterThan(before);
+      });
+
+      it("provides a minor increase in rank", () => {
+        basicStats(), ({ rank: before } = inst), complete(Field);
+        expect(inst.rank).toBeGreaterThan(before);
+      });
+    });
+
+    describe.each([SampleContract, SampleOperation, BlackOperations["Operation Archangel"].id])(
+      "non-general actions increase rank",
+      (id) => {
+        it(`${id.type}`, () => {
+          guaranteeSuccess(), (before = inst.rank), complete(id);
+          expect(inst.rank).toBeGreaterThan(before);
+        });
+      },
+    );
+
+    describe("non-general actions increase rank", () => {
+      let beforeMinor, minorGain, beforeMajor, majorGain;
+
+      it.each([
+        { major: SampleBlackOp, minor: SampleOperation },
+        { major: SampleOperation, minor: SampleContract },
+      ])("$major.type reward significantly more rank than $minor.type", ({ major, minor }) => {
+        guaranteeSuccess(), (beforeMinor = inst.rank), complete(minor), finish(), (minorGain = inst.rank - beforeMinor);
+        guaranteeSuccess(), (beforeMajor = inst.rank), complete(major), finish(), (majorGain = inst.rank - beforeMajor);
+        expect(majorGain).toBeGreaterThan(minorGain);
+      });
+    });
+
+    describe(BladeburnerGeneralActionName.InciteViolence, () => {
+      const Incite = GeneralAction.createId(BladeburnerGeneralActionName.InciteViolence);
+      let chaos;
+
+      it("generates available contracts", () => {
+        const { count } = inst.getActionObject(SampleContract);
+        guaranteeSuccess(), complete(Incite);
+        expect(inst.getActionObject(SampleContract).count).toBeGreaterThan(count);
+      });
+
+      it("generates available operations", () => {
+        const { count } = inst.getActionObject(SampleOperation);
+        guaranteeSuccess(), complete(Incite);
+        expect(inst.getActionObject(SampleOperation).count).toBeGreaterThan(count);
+      });
+
+      /** Relates to all issues mentioned in PR-1586 */
```

Comment:
Write out the issues here, it shouldn't require cross-referencing to understand what's being talked about.